File: bindings/ruby/lib/antech_kdf.rb
# frozen_string_literal: true

require "ffi"
require "rbconfig"

# Thin Ruby FFI wrapper (bindings/c/antech_kdf.h).

module Antech
  VERSION = "0.1.0"
  GRAPH_COMBINED_FRONTIER = 3

  class Error < StandardError; end

  Config = Struct.new(
    :memory_kib, :salt_length, :block_size, :fan_in, :graph, :output_length,
    keyword_init: true
  ) do
    def self.default
      c = Native::AntechConfig.new
      Native.raise!(Native.antech_config_default(c))
      new(
        memory_kib: c[:memory_kib],
        salt_length: c[:salt_length],
        block_size: c[:block_size],
        fan_in: c[:fan_in],
        graph: c[:graph],
        output_length: c[:output_length]
      )
    end

    def to_native
      c = Native::AntechConfig.new
      c[:memory_kib] = memory_kib
      c[:salt_length] = salt_length
      c[:block_size] = block_size
      c[:fan_in] = fan_in
      c[:graph] = graph
      c[:output_length] = output_length
      c
    end
  end

  RehashPolicy = Struct.new(
    :minimum_memory_kib, :preferred_memory_kib, :preferred_fan_in,
    :preferred_output_length, :preferred_secret_required, :preferred_associated_data,
    keyword_init: true
  ) do
    def self.default
      p = Native::AntechRehashPolicy.new
      Native.raise!(Native.antech_rehash_policy_default(p))
      new(
        minimum_memory_kib: p[:minimum_memory_kib],
        preferred_memory_kib: p[:preferred_memory_kib],
        preferred_fan_in: p[:preferred_fan_in],
        preferred_output_length: p[:preferred_output_length],
        preferred_secret_required: p[:preferred_secret_required] != 0,
        preferred_associated_data: p[:preferred_associated_data] != 0
      )
    end

    def to_native
      p = Native::AntechRehashPolicy.new
      p[:minimum_memory_kib] = minimum_memory_kib
      p[:preferred_memory_kib] = preferred_memory_kib
      p[:preferred_fan_in] = preferred_fan_in
      p[:preferred_output_length] = preferred_output_length
      p[:preferred_secret_required] = preferred_secret_required ? 1 : 0
      p[:preferred_associated_data] = preferred_associated_data ? 1 : 0
      p
    end
  end

  module Native
    extend FFI::Library

    class AntechConfig < FFI::Struct
      layout :memory_kib, :uint32,
             :salt_length, :uint32,
             :block_size, :uint32,
             :fan_in, :uint32,
             :graph, :uint32,
             :output_length, :uint32
    end

    class AntechRehashPolicy < FFI::Struct
      layout :minimum_memory_kib, :uint32,
             :preferred_memory_kib, :uint32,
             :preferred_fan_in, :uint32,
             :preferred_output_length, :uint32,
             :preferred_secret_required, :uint32,
             :preferred_associated_data, :uint32
    end

    def self.library_path
      env = ENV["ANTECH_KDF_LIB"]
      root = File.expand_path("../../..", __dir__)
      names =
        case RbConfig::CONFIG["host_os"]
        when /mswin|mingw|cygwin/i then %w[antech_kdf.dll antech_kdf_ffi.dll]
        when /darwin/i then %w[libantech_kdf.dylib libantech_kdf_ffi.dylib]
        else %w[libantech_kdf.so libantech_kdf_ffi.so]
        end
      dirs = [
        (File.file?(env.to_s) ? File.dirname(env) : env),
        File.join(root, "sdk", "native"),
        File.join(root, "target", "release"),
        File.join(root, "target", "debug"),
        File.join(__dir__, "native")
      ].compact.reject(&:empty?)
      return env if env && File.file?(env)

      dirs.each do |d|
        names.each do |n|
          p = File.join(d, n)
          return p if File.file?(p)
        end
      end
      raise Error, "native library not found; run sdk/scripts/build-native.(sh|ps1) or set ANTECH_KDF_LIB"
    end

    ffi_lib library_path

    attach_function :antech_version, [], :string
    attach_function :antech_free, [:pointer], :void
    attach_function :antech_config_default, [AntechConfig.by_ref], :int
    attach_function :antech_rehash_policy_default, [AntechRehashPolicy.by_ref], :int
    attach_function :antech_hash_bytes, [:pointer, :size_t, :pointer], :int
    attach_function :antech_hash_with_config_bytes, [:pointer, :size_t, AntechConfig.by_ref, :pointer], :int
    attach_function :antech_hash_with_config_and_salt,
                    [:pointer, :size_t, :pointer, :size_t, AntechConfig.by_ref, :pointer], :int
    attach_function :antech_hash_with_inputs_bytes,
                    [:pointer, :size_t, AntechConfig.by_ref, :pointer, :size_t, :pointer, :size_t, :pointer], :int
    attach_function :antech_hash_with_inputs_and_salt,
                    [:pointer, :size_t, :pointer, :size_t, AntechConfig.by_ref, :pointer, :size_t, :pointer, :size_t, :pointer], :int
    attach_function :antech_verify_bytes, [:pointer, :size_t, :string], :int
    attach_function :antech_verify_with_inputs_bytes,
                    [:pointer, :size_t, :string, :pointer, :size_t, :pointer, :size_t], :int
    attach_function :antech_needs_rehash, [:string, :pointer], :int
    attach_function :antech_needs_rehash_with_policy, [:string, AntechRehashPolicy.by_ref, :pointer], :int

    module_function

    def raise!(st)
      return if st == 0

      msg =
        case st
        when -1 then "invalid input"
        when -2 then "invalid hash"
        when -4 then "invalid config"
        else "internal error (#{st})"
        end
      raise Error, msg
    end

    def take(out_ptr)
      p = out_ptr.read_pointer
      raise Error, "null string" if p.null?

      s = p.read_string
      antech_free(p)
      s
    end

    # nil = absent; empty string = present empty.
    def opt_buf(data)
      return [FFI::Pointer::NULL, 0] if data.nil?
      return [FFI::MemoryPointer.new(:uint8, 1), 0] if data.bytesize.zero?

      buf = FFI::MemoryPointer.new(:uint8, data.bytesize)
      buf.put_bytes(0, data)
      [buf, data.bytesize]
    end

    def bytes_buf(data)
      return [FFI::Pointer::NULL, 0] if data.nil? || data.bytesize.zero?

      buf = FFI::MemoryPointer.new(:uint8, data.bytesize)
      buf.put_bytes(0, data)
      [buf, data.bytesize]
    end
  end

  module_function

  def version
    Native.antech_version || VERSION
  end

  def hash(password)
    pw, len = Native.bytes_buf(password.to_s)
    out = FFI::MemoryPointer.new(:pointer)
    Native.raise!(Native.antech_hash_bytes(pw, len, out))
    Native.take(out)
  end

  def hash_with_config(password, config)
    pw, len = Native.bytes_buf(password.to_s)
    out = FFI::MemoryPointer.new(:pointer)
    Native.raise!(Native.antech_hash_with_config_bytes(pw, len, config.to_native, out))
    Native.take(out)
  end

  def hash_with_config_and_salt(password, salt, config)
    pw, pw_len = Native.bytes_buf(password.to_s)
    s, s_len = Native.bytes_buf(salt.to_s)
    out = FFI::MemoryPointer.new(:pointer)
    Native.raise!(Native.antech_hash_with_config_and_salt(pw, pw_len, s, s_len, config.to_native, out))
    Native.take(out)
  end

  def hash_with_inputs(password, config, secret: nil, associated_data: nil)
    pw, pw_len = Native.bytes_buf(password.to_s)
    sec, sec_len = Native.opt_buf(secret)
    ad, ad_len = Native.opt_buf(associated_data)
    out = FFI::MemoryPointer.new(:pointer)
    Native.raise!(Native.antech_hash_with_inputs_bytes(
      pw, pw_len, config.to_native, sec, sec_len, ad, ad_len, out
    ))
    Native.take(out)
  end

  def hash_with_inputs_and_salt(password, salt, config, secret: nil, associated_data: nil)
    pw, pw_len = Native.bytes_buf(password.to_s)
    s, s_len = Native.bytes_buf(salt.to_s)
    sec, sec_len = Native.opt_buf(secret)
    ad, ad_len = Native.opt_buf(associated_data)
    out = FFI::MemoryPointer.new(:pointer)
    Native.raise!(Native.antech_hash_with_inputs_and_salt(
      pw, pw_len, s, s_len, config.to_native, sec, sec_len, ad, ad_len, out
    ))
    Native.take(out)
  end

  def verify(password, encoded_hash)
    pw, len = Native.bytes_buf(password.to_s)
    st = Native.antech_verify_bytes(pw, len, encoded_hash)
    return true if st == 0
    return false if st == 1

    Native.raise!(st)
  end

  def verify_with_inputs(password, encoded_hash, secret: nil, associated_data: nil)
    pw, len = Native.bytes_buf(password.to_s)
    sec, sec_len = Native.opt_buf(secret)
    ad, ad_len = Native.opt_buf(associated_data)
    st = Native.antech_verify_with_inputs_bytes(pw, len, encoded_hash, sec, sec_len, ad, ad_len)
    return true if st == 0
    return false if st == 1

    Native.raise!(st)
  end

  def needs_rehash(encoded_hash)
    out = FFI::MemoryPointer.new(:int)
    Native.raise!(Native.antech_needs_rehash(encoded_hash, out))
    out.read_int != 0
  end

  def needs_rehash_with_policy(encoded_hash, policy)
    out = FFI::MemoryPointer.new(:int)
    Native.raise!(Native.antech_needs_rehash_with_policy(encoded_hash, policy.to_native, out))
    out.read_int != 0
  end
end

File: bindings/lua/antech_kdf.lua
-- Thin LuaJIT FFI wrapper (bindings/c/antech_kdf.h).

local ffi = require("ffi")

ffi.cdef[[
typedef struct AntechConfig {
  uint32_t memory_kib;
  uint32_t salt_length;
  uint32_t block_size;
  uint32_t fan_in;
  uint32_t graph;
  uint32_t output_length;
} AntechConfig;

typedef struct AntechRehashPolicy {
  uint32_t minimum_memory_kib;
  uint32_t preferred_memory_kib;
  uint32_t preferred_fan_in;
  uint32_t preferred_output_length;
  uint32_t preferred_secret_required;
  uint32_t preferred_associated_data;
} AntechRehashPolicy;

const char* antech_version(void);
int antech_config_default(AntechConfig* out);
int antech_rehash_policy_default(AntechRehashPolicy* out);
int antech_hash_bytes(const uint8_t* password, size_t password_len, char** out_hash);
int antech_hash_with_config_bytes(const uint8_t* password, size_t password_len, const AntechConfig* config, char** out_hash);
int antech_hash_with_config_and_salt(const uint8_t* password, size_t password_len, const uint8_t* salt, size_t salt_len, const AntechConfig* config, char** out_hash);
int antech_hash_with_inputs_bytes(const uint8_t* password, size_t password_len, const AntechConfig* config, const uint8_t* secret, size_t secret_len, const uint8_t* associated_data, size_t associated_data_len, char** out_hash);
int antech_hash_with_inputs_and_salt(const uint8_t* password, size_t password_len, const uint8_t* salt, size_t salt_len, const AntechConfig* config, const uint8_t* secret, size_t secret_len, const uint8_t* associated_data, size_t associated_data_len, char** out_hash);
int antech_verify_bytes(const uint8_t* password, size_t password_len, const char* encoded_hash);
int antech_verify_with_inputs_bytes(const uint8_t* password, size_t password_len, const char* encoded_hash, const uint8_t* secret, size_t secret_len, const uint8_t* associated_data, size_t associated_data_len);
int antech_needs_rehash(const char* encoded_hash, int* out_needs_rehash);
int antech_needs_rehash_with_policy(const char* encoded_hash, const AntechRehashPolicy* policy, int* out_needs_rehash);
void antech_free(char* ptr);
]]

local M = {
  VERSION = "0.1.0",
  GRAPH_COMBINED_FRONTIER = 3,
}

local lib

local function is_windows()
  return package.config:sub(1, 1) == "\\"
end

local function library_names()
  if is_windows() then
    return { "antech_kdf.dll", "antech_kdf_ffi.dll" }
  end
  local uname = (io.popen and io.popen("uname -s"):read("*l")) or ""
  if uname == "Darwin" then
    return { "libantech_kdf.dylib", "libantech_kdf_ffi.dylib" }
  end
  return { "libantech_kdf.so", "libantech_kdf_ffi.so" }
end

local function file_exists(path)
  local f = io.open(path, "rb")
  if f then f:close() return true end
  return false
end

local function join(...)
  local sep = is_windows() and "\\" or "/"
  return table.concat({ ... }, sep)
end

function M.find_library()
  local env = os.getenv("ANTECH_KDF_LIB")
  if env and file_exists(env) then return env end
  local here = debug.getinfo(1, "S").source:sub(2)
  local root = here:match("^(.*)[/\\]bindings[/\\]lua")
  local dirs = {
    env,
    root and join(root, "sdk", "native") or nil,
    root and join(root, "target", "release") or nil,
    root and join(root, "target", "debug") or nil,
    join("sdk", "native"),
    join("target", "release"),
  }
  for _, d in ipairs(dirs) do
    if d then
      for _, n in ipairs(library_names()) do
        local p = join(d, n)
        if file_exists(p) then return p end
      end
    end
  end
  error("native library not found; run sdk/scripts/build-native.(sh|ps1) or set ANTECH_KDF_LIB")
end

local function ensure()
  if lib then return lib end
  lib = ffi.load(M.find_library())
  return lib
end

local function raise(st)
  if st == 0 then return end
  local msg = ({
    [-1] = "invalid input",
    [-2] = "invalid hash",
    [-4] = "invalid config",
  })[st] or ("internal error (" .. tostring(st) .. ")")
  error(msg)
end

local function take(out_ptr)
  local p = out_ptr[0]
  if p == nil then error("null string") end
  local s = ffi.string(p)
  ensure().antech_free(p)
  return s
end

-- nil = absent; "" = present empty
local function opt_buf(data)
  if data == nil then return nil, 0 end
  if #data == 0 then
    local buf = ffi.new("uint8_t[1]")
    return buf, 0
  end
  local buf = ffi.new("uint8_t[?]", #data)
  ffi.copy(buf, data, #data)
  return buf, #data
end

local function bytes_buf(data)
  data = data or ""
  if #data == 0 then
    local buf = ffi.new("uint8_t[1]")
    return buf, 0
  end
  local buf = ffi.new("uint8_t[?]", #data)
  ffi.copy(buf, data, #data)
  return buf, #data
end

function M.version()
  local v = ensure().antech_version()
  return v ~= nil and ffi.string(v) or M.VERSION
end

function M.config_default()
  local c = ffi.new("AntechConfig")
  raise(ensure().antech_config_default(c))
  return {
    memory_kib = tonumber(c.memory_kib),
    salt_length = tonumber(c.salt_length),
    block_size = tonumber(c.block_size),
    fan_in = tonumber(c.fan_in),
    graph = tonumber(c.graph),
    output_length = tonumber(c.output_length),
  }
end

local function to_c_config(cfg)
  local c = ffi.new("AntechConfig")
  c.memory_kib = cfg.memory_kib
  c.salt_length = cfg.salt_length
  c.block_size = cfg.block_size
  c.fan_in = cfg.fan_in
  c.graph = cfg.graph
  c.output_length = cfg.output_length
  return c
end

function M.hash(password)
  local pw, len = bytes_buf(password)
  local out = ffi.new("char*[1]")
  raise(ensure().antech_hash_bytes(pw, len, out))
  return take(out)
end

function M.hash_with_config(password, config)
  local pw, len = bytes_buf(password)
  local out = ffi.new("char*[1]")
  raise(ensure().antech_hash_with_config_bytes(pw, len, to_c_config(config), out))
  return take(out)
end

function M.hash_with_config_and_salt(password, salt, config)
  local pw, pw_len = bytes_buf(password)
  local s, s_len = bytes_buf(salt)
  local out = ffi.new("char*[1]")
  raise(ensure().antech_hash_with_config_and_salt(pw, pw_len, s, s_len, to_c_config(config), out))
  return take(out)
end

function M.hash_with_inputs(password, config, secret, associated_data)
  local pw, pw_len = bytes_buf(password)
  local sec, sec_len = opt_buf(secret)
  local ad, ad_len = opt_buf(associated_data)
  local out = ffi.new("char*[1]")
  raise(ensure().antech_hash_with_inputs_bytes(
    pw, pw_len, to_c_config(config), sec, sec_len, ad, ad_len, out))
  return take(out)
end

function M.verify(password, encoded_hash)
  local pw, len = bytes_buf(password)
  local st = ensure().antech_verify_bytes(pw, len, encoded_hash)
  if st == 0 then return true end
  if st == 1 then return false end
  raise(st)
end

function M.verify_with_inputs(password, encoded_hash, secret, associated_data)
  local pw, len = bytes_buf(password)
  local sec, sec_len = opt_buf(secret)
  local ad, ad_len = opt_buf(associated_data)
  local st = ensure().antech_verify_with_inputs_bytes(
    pw, len, encoded_hash, sec, sec_len, ad, ad_len)
  if st == 0 then return true end
  if st == 1 then return false end
  raise(st)
end

function M.needs_rehash(encoded_hash)
  local out = ffi.new("int[1]")
  raise(ensure().antech_needs_rehash(encoded_hash, out))
  return out[0] ~= 0
end

return M
